src/target/ch579.c

```diff
@@ -0,0 +1,139 @@
+{
+    (void)len;
+    DEBUG_INFO("ch579 flash erase %08x\n", (int)addr);
+    target_mem_write32(f->t, 0x40001804, addr);
+    target_mem_write8(f->t, 0x40001808, 0xa6);
+
+    if (!ch579_wait_flash(f->t, NULL))
+        return false;
+    return true;
+}
+static bool ch579_flash_write(target_flash_s *f, target_addr_t dest, const void *src, size_t len)
+{
+    (void)len;
+    DEBUG_INFO("ch579 flash write %08x\n", (int)dest);
+    target_mem_write32(f->t, 0x40001804, dest);
+    target_mem_write32(f->t, 0x40001800, read_le4((const uint8_t *)src, 0));
+    target_mem_write8(f->t, 0x40001808, 0x9a);
+
+    if (!ch579_wait_flash(f->t, NULL))
+        return false;
+    return true;
+}
+static bool ch579_flash_prepare(target_flash_s *f)
+{
+    DEBUG_INFO("ch579 flash prepare\n");
+    // just enable both write flags now, so that code/data flash can be treated as contiguous
+    target_mem_write8(f->t, 0x40001809, 0b10001100);
+    return true;
+}
+static bool ch579_flash_done(target_flash_s *f)
+{
+    DEBUG_INFO("ch579 flash done\n");
+    target_mem_write8(f->t, 0x40001809, 0b10000000);
+    return true;
+}
+
+static bool ch579_cmd_erase_info_dangerous(target_s *target, int argc, const char **argv)
+{
+    (void)argc;
+    (void)argv;
+    target_mem_write8(target, 0x40001809, 0b10001100);
+    target_mem_write32(target, 0x40001804, 0x40000);
+    target_mem_write8(target, 0x40001808, 0xa5);
+    bool okay = ch579_wait_flash(target, NULL);
+    target_mem_write8(target, 0x40001809, 0b10000000);
+    return okay;
+}
+static bool ch579_cmd_write_info_dangerous(target_s *target, int argc, const char **argv)
+{
+	uint32_t addr;
+	uint32_t val;
+
+	if (argc == 3) {
+		addr = strtoul(argv[1], NULL, 0);
+		val = strtoul(argv[2], NULL, 0);
+	} else
+		return false;
+
+    target_mem_write8(target, 0x40001809, 0b10001100);
+    target_mem_write32(target, 0x40001804, addr);
+    target_mem_write32(target, 0x40001800, val);
+    target_mem_write8(target, 0x40001808, 0x99);
+    bool okay = ch579_wait_flash(target, NULL);
+    target_mem_write8(target, 0x40001809, 0b10000000);
+    return okay;
+}
+static bool ch579_cmd_disable_bootloader(target_s *target, int argc, const char **argv)
+{
+    (void)argc;
+    (void)argv;
+    target_mem_write8(target, 0x40001809, 0b10001100);
+    target_mem_write32(target, 0x40001804, 0x40010);
+    target_mem_write32(target, 0x40001800, 0xFFFFFFBF);
+    target_mem_write8(target, 0x40001808, 0x99);
+    bool okay = ch579_wait_flash(target, NULL);
+    target_mem_write8(target, 0x40001809, 0b10000000);
+    return okay;
+}
+
+const command_s ch579_cmd_list[] = {
+	{"void_warranty_erase_infoflash", ch579_cmd_erase_info_dangerous, "Erase info flash sector"},
+	{"void_warranty_write_infoflash", ch579_cmd_write_info_dangerous, "Write to info flash: [address] [value]"},
+    {"disable_bootloader", ch579_cmd_disable_bootloader, "Disables ISP bootloader"},
+	{NULL, NULL, NULL},
+};
+
+
+bool ch579_probe(target_s *target)
```

Comment:
We know that you've put this all the way down here to save on function prototypes, but please don't.

We keep referencing the STM32H5 support as it provides a clear guide on the style we expect from target support - the prototypes exist so the code gets ordered more logically: probe handling, Flash support, monitor commands. It also provides consistency between the different target support implementations.

Likewise, it demonstrates the define prefixing and naming scheme so we can have named addresses and bit values rather than undocumented magic numbers everywhere for the next person who has to touch this code, which could be yourself in 6mo after you've forgotten what it all means.